keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs/README.md

```diff
@@ -0,0 +1,1430 @@
+(see in [code](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/job/backoff_utils.go#L160-L182)).
+Thus, there is a risk that due to the presence
+of pods hitting the fallback the last failure time is continuously bumped,
+thus shifting the time to recreate the pod.
+
+This risk is present both when computing the expotential backoff delay globally
+(as for regular indexed Jobs), or per-index as proposed in in this KEP
+(see [Expotential backoff delay per index](#expotential-backoff-delay-per-index)).
+
+In order to mitigate this risk currently the time of last failure is recorded
+in-memory (globally for all pods within a Job). And a new failed pod may bump
+it only until it is added to the `uncountedTerminatedPods` structure.
+
+However, tracking the last failure time per index might be costly for memory
+consumption (see [Expotential backoff delay with in-memory tracking](#expotential-backoff-delay-with-in-memory-tracking)).
+
+Thus, in order to mitigate this risk we propose to compute the finish time for
+a pod as the first available value of the following (avoiding the ever-increasing
+fallback to `now`):
+1. max `finishAt` of all containers, if specified for all containers
+2. `LastTransitionTime` for the `Ready=False` condition
+3. `deletionTimestamp` - `deletionGracePeriodSeconds` if `deletionTimestamp` is set
+
+Here (3.) is used to mark the moment of deletion which is used to approximate
+the current behavior. (2.) is used when Kubelet loses track of one of its containers,
+the `Ready=False` condition is set by Kubelet when transitioning a pod to `Failed`
+phase: https://github.com/kubernetes/kubernetes/blob/release-1.27/pkg/kubelet/status/status_manager.go#L1060-L1068.
+When none of the above conditions is satisfied to compute the finish time we
+fallback to the pod's creation time.
+
+This fix can be considered a preparatory PR before the KEP, as to some extent
+is solves the preexisting issue.
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+We introduce a new Job API field, called `.spec.backoffLimitPerIndex`.
+When set it limits the number of retries, counted independently for all indexes.
+
+Additionally, we propose the `.spec.maxFailedIndexes` to control
+the maximal number of failed indexes. Once the number is exceeded the entire
+Job is marked Failed and its execution is terminated.
+
+We also propose to extend the PodFailurePolicy with a new action, called
+`FailIndex` to allow an index to fail fast before reaching the backoff limit
+per index.
+
+### Job API
+
+```golang
+
+// PodFailurePolicyAction specifies how a Pod failure is handled.
+// +enum
+type PodFailurePolicyAction string
```

Comment:
This only defines a type - can you add a field (I'm assuming to JobSpec) that will take advantage of it?

Reply:
we already have a field: https://github.com/kubernetes/kubernetes/blob/72a3990728b2a8979effb37b9800beb3117349f6/pkg/apis/batch/types.go#L204
It type is repeated in the KEP to set the context.

Reply:
Added a small note to clarify.

Reply:
OK - sorry, somehow missed that.